src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,288 @@
+    | Ana(ana_ty) => Some(Typ.normalize(ctx, ana_ty))
+    }
+  | _ => None
+  };
+};
+
+let ground = (ty: Typ.t): bool => {
+  switch (ty) {
+  | Bool
+  | Int
+  | Float
+  | String
+  | Prod([])
+  | Arrow(Unknown(Internal), Unknown(Internal)) => true
+  | _ => false
+  };
+};
+
+let rec dhpat_extend_ctx = (dhpat: DHPat.t, ty: Typ.t, ctx: Ctx.t): Ctx.t => {
+  switch (dhpat, ty) {
+  | (Var(name), _) =>
+    let entry = Ctx.VarEntry({name, id: Id.invalid, typ: ty});
+    Ctx.extend(ctx, entry);
+  | (Tuple(l1), Prod(l2)) =>
+    if (List.length(l1) == List.length(l2)) {
+      List.fold_left2(
+        (acc, dhp, typ) => {dhpat_extend_ctx(dhp, typ, acc)},
+        ctx,
+        l1,
+        l2,
+      );
+    } else {
+      ctx;
+    }
+  | (Cons(dhp1, dhp2), List(typ)) =>
+    ctx |> dhpat_extend_ctx(dhp1, typ) |> dhpat_extend_ctx(dhp2, ty)
+  | (ListLit(typ1, l), List(typ2)) =>
+    if (Typ.eq(typ1, typ2)) {
+      List.fold_left(
+        (acc, dhp) => {dhpat_extend_ctx(dhp, typ1, acc)},
+        ctx,
+        l,
+      );
+    } else {
+      ctx;
+    }
+  | (Ap(Constructor(_, typ), dhp), _) =>
+    let (ty1, ty2) = Typ.matched_arrow(ctx, typ);
+    if (Typ.eq(ty2, ty)) {
+      ctx |> dhpat_extend_ctx(dhp, ty1);
+    } else {
+      ctx;
+    };
+  | _ => ctx
+  };
+};
+
+let rec typ_of_dhexp =
+        (ctx: Ctx.t, m: Statics.Map.t, dh: DHExp.t): option(Typ.t) => {
+  switch (dh) {
+  | EmptyHole(id, _) => delta_ty(id, m)
+  | NonEmptyHole(_, id, _, d) =>
+    switch (typ_of_dhexp(ctx, m, d)) {
+    | None => None
+    | Some(_) => delta_ty(id, m)
+    }
+  | ExpandingKeyword(_) => None
+  | FreeVar(_) => Some(Unknown(Internal))
+  | InvalidText(_) => None
+  | InconsistentBranches(_, _, Case(d_scrut, d_rules, _)) =>
+    let* ty' = typ_of_dhexp(ctx, m, d_scrut);
+    let typ_cases =
+      d_rules
+      |> List.map(rule_prj)
+      |> List.map(((dhp, de)) => {
+           typ_of_dhexp(dhpat_extend_ctx(dhp, ty', ctx), m, de)
+         })
+      |> OptUtil.sequence;
+
+    switch (typ_cases) {
+    | None => None
+    | Some(_) => Some(Typ.Unknown(Internal))
+    };
+  | Closure(_, d) => typ_of_dhexp(ctx, m, d)
+  | Filter(_, d) => typ_of_dhexp(ctx, m, d)
+  | BoundVar(name) =>
+    let+ var = Ctx.lookup_var(ctx, name);
+    var.typ;
+  | Sequence(d1, d2) =>
+    let* _ = typ_of_dhexp(ctx, m, d1);
+    typ_of_dhexp(ctx, m, d2);
+  | Let(dhp, de, db) =>
+    let* ty1 = typ_of_dhexp(ctx, m, de);
+    typ_of_dhexp(dhpat_extend_ctx(dhp, ty1, ctx), m, db);
+  | FixF(name, ty1, d) =>
+    let entry = Ctx.VarEntry({name, id: Id.invalid, typ: ty1});
+    typ_of_dhexp(Ctx.extend(ctx, entry), m, d);
+  | Fun(dhp, ty1, d, _) =>
+    let+ ty2 = typ_of_dhexp(dhpat_extend_ctx(dhp, ty1, ctx), m, d);
+    Typ.Arrow(ty1, ty2);
+  | Ap(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (arrow_aux(ty1)) {
+    | Arrow(tyl, tyr) when Typ.eq(tyl, ty2) => Some(tyr)
```

Comment:
we don't do matched arrow types in the internal language -- that should all be done via casts

Reply:
this is still an issue here